Common/Algorithm/Framework/Alphas/Analysis/InsightManager.cs

```diff
@@ -255,14 +257,21 @@ private void UpdateScores(ReadOnlySecurityValuesCollection securityValuesCollect
                     var id = context.Insight.Id;
                     _closedInsightContexts[id] = context;
 
+                    if (removals == null)
+                    {
+                        removals = new List<InsightAnalysisContext>();
+                    }
                     removals.Add(context);
                 }
 
                 // mark the context as having been updated
                 _updatedInsightContexts.Add(context);
             }
 
-            _openInsightContexts.RemoveWhere(removals.Contains);
+            if (removals != null)
+            {
+                _openInsightContexts.RemoveWhere(removals.Contains);
```

Comment:
For large numbers of insights, we might benefit from using a `HashSet<T>` here for `removals` to provide the `O(1)` `Contains` call.